builder/builder-next/exporter/wrapper.go

```diff
@@ -4,26 +4,39 @@ import (
 	"context"
 	"strings"
 
+	"github.com/containerd/log"
+	"github.com/distribution/reference"
 	"github.com/docker/docker/builder/builder-next/exporter/overrides"
 	"github.com/moby/buildkit/exporter"
 	"github.com/moby/buildkit/exporter/containerimage/exptypes"
 
 	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
 )
 
-type ImageExportedByBuildkit = func(ctx context.Context, id string, desc ocispec.Descriptor)
+type (
+	ImageExportedByBuildkit = func(ctx context.Context, id string, desc ocispec.Descriptor)
+	ImageNamedByBuildkit    = func(ctx context.Context, ref reference.NamedTagged, desc ocispec.Descriptor)
+)
+
+type BuildkitCallbacks struct {
+	Exported ImageExportedByBuildkit
+	Named    ImageNamedByBuildkit
+}
```

Comment:
What if we would change the signature of the callbacks to take `event.Action` as argument? In that case the callback could decide whether it needs to do anything.

We could either change the `calllbacks` to `[]Callback` or (perhaps less code-churn) continue to accept a single callback (which could either have a `switch event.Action`, or be a wrapper for multiple callbacks)

Reply:
I don't like that because I consider the `event.Action` to be an implementation detail here.
These callbacks are meant to provide the integration point between the image service/daemon and the builder, that was naturally present with the legacy builder and is lost when using Buildkit.
This can potentially be useful for other things, not only emitting events, so IMO it doesn't make sense for the builder code to be concerned with daemon events at all.